src/pveclib/vec_int128_ppc.h

```diff
@@ -903,6 +1059,107 @@ example_longdiv_10e31 (vui128_t *q, vui128_t *d, long int _N)
+    +    ((unsigned __int128) __q1)) << 32) \
+    +    ((unsigned __int128) __q2)) << 32) \
+    +    ((unsigned __int128) __q3) )
+
+/** \brief Generate a vector unsigned __int128 constant from doublewords.
+ *
+ *  Combine 2 x 64-bit long long constants into a single __int128 constant.
+ *  The 2 parameters are long integer constant values in high to low order.
+ *  This order is consistent for big and little endian and the result
+ *  loaded into vector registers is correct for quadword  integer
+ *  operations.
+ *
+ *  For example
+ *
+ * \code
+ vui128_t ten32 = CONST_VUINT128_QxD (0x000004ee2d6d415bUL, 0x85acef8100000000UL);
+ * \endcode
+ *
+ */
+#define CONST_VUINT128_QxD(__q0, __q1) ( (vui128_t) \
+    (((unsigned __int128) __q0) << 64) \
+    + (((unsigned __int128) __q1) )
+
+/** \brief Generate a vector unsigned __int128 constant from doublewords.
+ *
+ *  Combine 2 x 19 decimal digit long long constants into a single
+ *  38 decimal digit __int128 constant.
+ *  The 2 parameters are long integer constant values in high to low order.
+ *  This order is consistent for big and little endian and the result
+ *  loaded into vector registers is correct for quadword  integer
+ *  operations.
+ *
+ *  For example
+ *
+ * \code
+ const vui128_t mul_invs_ten16 = CONST_VUINT128_Qx19d(
+      7662477704329444291UL, 7917351357515459181UL);
+ * \endcode
+ *
+ */
+#define CONST_VUINT128_Qx19d(__q0, __q1) ( (vui128_t) \
+    (((unsigned __int128) __q0) * 10000000000000000000UL) \
+    + ((unsigned __int128) __q1) )
+
+/** \brief Generate a vector unsigned __int128 constant from doublewords.
+ *
+ *  Combine 2 x 18 decimal digit long long constants into a single
+ *  36 decimal digit __int128 constant.
+ *  The 2 parameters are long integer constant values in high to low order.
+ *  This order is consistent for big and little endian and the result
+ *  loaded into vector registers is correct for quadword  integer
+ *  operations.
+ *
+ *  For example
+ *
+ * \code
+ vui128_t ten36-1 = CONST_VUINT128_Qx18d (999999999999999999UL, 999999999999999999UL);
+ * \endcode
+ *
+ */
+#define CONST_VUINT128_Qx18d(__q0, __q1) ( (vui128_t) \
+    (((unsigned __int128) __q0) * 1000000000000000000UL) \
+    + ((unsigned __int128) __q1) )
+
+/** \brief Generate a vector unsigned __int128 constant from doublewords.
+ *
+ *  Combine 2 x 16 decimal digit long long constants into a single
+ *  32 decimal digit __int128 constant.
+ *  The 2 parameters are long integer constant values in high to low order.
+ *  This order is consistent for big and little endian and the result
+ *  loaded into vector registers is correct for quadword  integer
+ *  operations.
+ *
+ *  For example
+ *
+ * \code
+ const vui128_t ten32 = CONST_VUINT128_Qx16d (10000000000000000UL, 0UL);
+ * \endcode
+ *
+ */
+#define CONST_VUINT128_Qx16d(__q0, __q1) ( (vui128_t) \
+    (((unsigned __int128) __q0) * 10000000000000000UL) \
+    + ((unsigned __int128) __q1) )
+
```

Comment:
`10^n` for `n=19` is the largest `n` such that `10^n < 2^64`, so there's no boundary issues at 64-bit literals with the `UL` qualifier. The same holds for `n=38=2*19` and `2^128`. So you can safely construct a 128-bit literal from two 19-decimal digit chunks (since you can't have a single 38-decimal-digit literal due to the GCC restrictions mentioned above). And a single 39-decimal-digit constant exceeds the size constraint of `2^128`.

I do kinda agree that having separate 16, 18, and 19 digit macros are a bit excessive, other than making the resulting code pretty because you can break the literals into two even halves.

If you don't have an equal number of digits on each side, you have to then count all 19 digits out and make sure any future changes to those constants preserves the constraint that 19 are in the least significant place. I'm not sure there's any way of enforcing that constraint in a nice manner (sans, creating more macros...).

Reply:
I explained this a email.